### PR TITLE
test/upgrade: Add a single node upgrade scenario

### PR DESCRIPTION
**Pull request by Ebaneck**
Fixes: #1698

**Component**:

'upgrade', 'downgrade', 'tests'

**Context**: 

Introduce MetalK8s upgrade testing in the CI.

**Summary**:

 We should be able to run an upgrade from minor version N-1 to version N in the CI

**Acceptance criteria**: 

The upgrade scenario should be successful.

---

Closes: #ISSUE_NUMBER

### CODE REVIEW

File: eve/main.yml [hunk 1]
```diff
@@ -285,6 +285,134 @@ stages:
...
+      - ShellCommand:
+          name: Retrieve ISO image checksum
+          command: >
+            curl -s -XGET -o /tmp/SHA256SUM
+            "%(prop:artifacts_private_url)s/SHA256SUM"
+          haltOnFailure: true
+      - ShellCommand:
+          name: Check image with checksum
+          command: cd /tmp && sha256sum -c SHA256SUM
+          haltOnFailure: true
```
Comment (TeddyAndrieux) [inline eve/main.yml:362]: IMO you should retrieve and check both ISO at the beginning and use yaml tags see https://github.com/scality/metalk8s/pull/1730/files#r326499412

File: eve/main.yml [hunk 2]
```diff
@@ -285,6 +286,134 @@ stages:
...
+            --verbose || true
+          haltOnFailure: true
+      - ShellCommand:
+          name: Run upgrade to %(prop:metalk8s_version)s
+          command: >
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/iso-manager.sh
+            --archive /tmp/metalk8s.iso &&
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/upgrade.sh
```
Comment (TeddyAndrieux) [inline eve/main.yml]: No, please use the upgrade script of the newer version.

File: eve/main.yml [hunk 3]
```diff
@@ -285,6 +286,134 @@ stages:
...
+            curl -s -XGET -o /tmp/metalk8s.iso
+            "%(prop:artifacts_private_url)s/metalk8s.iso"
+          haltOnFailure: true
+      - ShellCommand: &get_iso_checksum
+          name: Get %(prop:metalk8s_version)s checksum
+          command: >
+            curl -s -XGET -o /tmp/SHA256SUM
+            "%(prop:artifacts_private_url)s/SHA256SUM"
+          haltOnFailure: true
+      - ShellCommand: &verify_iso_checksum
```
Comment (TeddyAndrieux) [inline eve/main.yml]: `&blabla` is special yaml Anchor (you can google it for more informations)
Here you redefine the `verify_iso_checksum`

You should remove all the content of hte ShellCommand and only keep 
```
- ShellCommand: *verify_iso_checksum
```

And btw it's the same for all other steps with `: &blabla`

And as suggest in downgrade PR you should add some for common upgrade/downgrade steps

Comment (Ebaneck) [reply eve/main.yml]: yes, we can do that. which ever PR gets merge first will re-use the shared tags. I got a green light from @sayf-eddine-scality  to go first if i didnt get a flaky.

Comment (TeddyAndrieux) [reply eve/main.yml]: Ok so lets definie tag in your PR like I suggested here
https://github.com/scality/metalk8s/pull/1730/files#r326615160

Comment (Ebaneck) [reply eve/main.yml]: ok!

File: eve/main.yml [hunk 4]
```diff
@@ -285,6 +286,144 @@ stages:
...
 
+  single-node-upgrade-test:
+    worker:
+      type: openstack
+      image: CentOS-7-x86_64-GenericCloud-1809.qcow2
+      flavor: m1.large
+      path: eve/workers/openstack-single-node
+    steps:
+      - Git: *git_pull
+      - ShellCommand:
```
Comment (TeddyAndrieux) [inline eve/main.yml]: Replace this block with 
```suggestion
      - ShellCommand: *ssh_ip_setup
```

File: eve/main.yml [hunk 5]
```diff
@@ -285,6 +286,144 @@ stages:
...
+          name: Create %(prop:metalk8s_prev_version)s mountpoint
+          command: >
+            sudo mkdir -p /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s
+      - ShellCommand: &mount_prev_iso
+          name: Mount %(prop:metalk8s_prev_version)s image
+          command: >
+            sudo mount -o loop metalk8s.iso
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s
+          haltOnFailure: true
+      - ShellCommand: &get_iso
```
Comment (TeddyAndrieux) [inline eve/main.yml]:
```suggestion
      - ShellCommand: *retrieve_iso
```

File: eve/main.yml [hunk 6]
```diff
@@ -285,6 +286,144 @@ stages:
...
+          name: Get %(prop:metalk8s_version)s image
+          command: >
+            curl -s -XGET -o /tmp/metalk8s.iso
+            "%(prop:artifacts_private_url)s/metalk8s.iso"
+          haltOnFailure: true
+      - ShellCommand: &get_iso_checksum
```
Comment (TeddyAndrieux) [inline eve/main.yml]:
```suggestion
      - ShellCommand: *retrieve_iso_checksum
```

File: eve/main.yml [hunk 7]
```diff
@@ -285,6 +286,144 @@ stages:
...
+      path: eve/workers/openstack-single-node
+    steps:
+      - Git: *git_pull
+      - ShellCommand:
+          name: Report IP and install SSH keys
+          command: >
+            ip a &&
+            mkdir -p ~/.ssh &&
+            echo "%(secret:ssh_pub_keys)s" >> ~/.ssh/authorized_keys
+      - SetPropertyFromCommand: &set_prev_version
```
Comment (TeddyAndrieux) [inline eve/main.yml:298]: I would prefer to define all this step in downgrade as downgrade will be the first step to run, so we can keep them here but declaration should be moved to downgrade step after

File: eve/main.yml [hunk 8]
```diff
@@ -285,6 +286,144 @@ stages:
...
+            curl -s -XGET -o /tmp/metalk8s.iso
+            "%(prop:artifacts_private_url)s/metalk8s.iso"
+          haltOnFailure: true
+      - ShellCommand: &get_iso_checksum
+          name: Get %(prop:metalk8s_version)s checksum
+          command: >
+            curl -s -XGET -o /tmp/SHA256SUM
+            "%(prop:artifacts_private_url)s/SHA256SUM"
+          haltOnFailure: true
+      - ShellCommand: &verify_iso_checksum
```
Comment (TeddyAndrieux) [inline eve/main.yml]:
```suggestion
      - ShellCommand: *check_iso_checksum
```

File: eve/main.yml [hunk 9]
```diff
@@ -285,6 +286,144 @@ stages:
...
+      - ShellCommand:
+          name: Collect logs using sosreport
+          command: >
+            sudo sosreport --all-logs
+            -o metalk8s -kmetalk8s.podlogs=True
+            -o containerd -kcontainerd.all=True -kcontainerd.logs=True
+            --batch --tmp-dir /var/tmp &&
+            sudo chown eve:eve /var/tmp/sosreport*
+          alwaysRun: true
+      - ShellCommand:
```
Comment (TeddyAndrieux) [inline eve/main.yml]: Please add a tag for "Collect logs using sosreport", "Copy sosreport" and "upload artifacts" in classic single-node and use the tag here

File: eve/main.yml [hunk 10]
```diff
@@ -285,6 +286,144 @@ stages:
...
+            "%(prop:artifacts_private_url)s/SHA256SUM"
+          haltOnFailure: true
+      - ShellCommand: &verify_iso_checksum
+          name: Verify %(prop:metalk8s_version)s checksum
+          command: cd /tmp && sha256sum -c SHA256SUM
+          haltOnFailure: true
+      - ShellCommand: &create_iso_mountpoint
+          name: Create %(prop:metalk8s_version)s mountpoint
+          command: >
+            sudo mkdir -p /srv/scality/metalk8s-%(prop:metalk8s_version)s
```
Comment (TeddyAndrieux) [inline eve/main.yml]: You do not need to create the directory

File: eve/main.yml [hunk 11]
```diff
@@ -285,6 +286,144 @@ stages:
...
+      - ShellCommand: &mount_iso
+          name: Mount %(prop:metalk8s_version)s image
+          command: >
+            sudo mount -o loop /tmp/metalk8s.iso
+            /srv/scality/metalk8s-%(prop:metalk8s_version)s
+          haltOnFailure: true
```
Comment (TeddyAndrieux) [inline eve/main.yml:333]: No do not mount it the iso-manager will do it

File: eve/main.yml [hunk 12]
```diff
@@ -285,6 +286,144 @@ stages:
...
+          command: >
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/bootstrap.sh
+            --verbose || true
+          haltOnFailure: true
+      - ShellCommand:
+          name: Run upgrade to %(prop:metalk8s_version)s
+          command: >
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_version)s/iso-manager.sh
```
Comment (TeddyAndrieux) [inline eve/main.yml]: you need to use the previous version here as normaly the new version is not available at this point

**Review by TeddyAndrieux — CHANGES_REQUESTED** [review]

File: eve/main.yml (hunk 7 above)
Comment (Ebaneck) [reply eve/main.yml:298]: yes, that makes absolute sense

File: eve/main.yml [hunk 13]
```diff
@@ -268,23 +269,112 @@ stages:
...
             mkdir -p sosreport/sosreport/single-node &&
             cp /var/tmp/sosreport* sosreport/sosreport/single-node
           alwaysRun: true
-      - Upload:
+      - Upload: &upload_report_artifact
           name: upload artifacts
           source: sosreport
           alwaysRun: true
-      - ShellCommand:
+      - ShellCommand: &resource_destroy_wait
```
Comment (TeddyAndrieux) [inline eve/main.yml]: Minor, but the name is not really good here IMO, it's a debug step and for this one maybe better to keep copy paste for the moment as we may want to active debug only for one particular step

File: eve/main.yml [hunk 14]
```diff
@@ -268,23 +269,112 @@ stages:
...
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/bootstrap.sh
+            --verbose || true
+          haltOnFailure: true
+      - ShellCommand:
+          name: Run upgrade to %(prop:metalk8s_version)s
+          command: >
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/iso-manager.sh
+            --archive metalk8s.iso &&
```
Comment (TeddyAndrieux) [inline eve/main.yml]: You need to give the full path here (maybe something we should fix in the script, but for the moment you need the full path)

File: eve/main.yml [hunk 15]
```diff
@@ -285,6 +286,100 @@ stages:
...
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/bootstrap.sh
+            --verbose || true
+          haltOnFailure: true
+      - ShellCommand:
+          name: Run upgrade to %(prop:metalk8s_version)s
+          command: >
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/iso-manager.sh
+            --archive $(readlink -f metalk8s.iso) &&
```
Comment (TeddyAndrieux) [inline eve/main.yml]:
```suggestion
            --archive "$(readlink -f metalk8s.iso)" &&
```

File: eve/main.yml [hunk 16]
```diff
@@ -285,6 +286,100 @@ stages:
...
+            archives:
+              - /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s
+            END
+            EOF
+      - ShellCommand:
+          name: Start %(prop:metalk8s_prev_version)s bootstrap
+          command: >
+            sudo bash
+            /srv/scality/metalk8s-%(prop:metalk8s_prev_version)s/bootstrap.sh
+            --verbose || true
```
Comment (TeddyAndrieux) [inline eve/main.yml]: Why `|| true` ?

**Review by TeddyAndrieux — APPROVED** [review]
LGTM, just waiting green builds

**Pull request merged**